WalletWasabi/WabiSabi/Crypto/CredentialRequesting/CredentialsRequest.cs

```diff
@@ -65,11 +64,47 @@ public abstract class CredentialsRequest
 		/// <summary>
 		/// Serial numbers used in the credential presentations.
 		/// </summary>
+		[JsonIgnore]
 		internal IEnumerable<GroupElement> SerialNumbers => Presented.Select(x => x.S);
 
 		/// <summary>
 		/// Indicates whether the message contains duplicated serial numbers or not.
 		/// </summary>
+		[JsonIgnore]
 		internal bool AreThereDuplicatedSerialNumbers => SerialNumbers.Distinct().Count() < SerialNumbers.Count();
+
+		public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```

Comment:
Because these checks are redundant with later validation in the credential issuer and proof system, and also partial, I think it's important to point this out (or maybe even rename the method?)

```suggestion
		/// <summary>
		/// Partially validate the request. This ensures it has the correct structure as a precondition for handling, but does not validate that the content of the request is valid which is done later.
		/// </summary>
		public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
```

Reply:
I am going to remove all this because it is part of the server-side contracts deserialization plumbing. ASP.NET allows us to implement this kind of validations at contract level to make sure to separate the inputs validation logic from the business logic. Anyway, this was too early.

Reply:
That sounds sensible. FWIW I do think this kind of checking is desirable, I think it's strictly better to reject invalid inputs earlier whenever possible, and the fact that it's redundant with the domain level validation is not a downside so long as both code paths are tested separately, as long as it's kept reasonably simple (like it is here) this kind of belt-and-suspenders approach helps in maintenance in my experience, and is probably also helpful for denial of service protection.

Reply:
Yes, this is good imo. I removed it for the moment.